### ISSUE TITLE
UDPROS does not work at armhf

### ISSUE THREAD

**Issue by AndrijCherepaha** [ros/ros_comm#1404]
The issue was found during experiments with RaspberryPi. Any attempts of subscribing to topics using TransportHints.udp() or TransportHints::unreliable() end with broken communication. There is no crashes on both subscriber's and publisher's sides.

Preconditions:
Install ubuntu to armhf (https://ubuntu-mate.org/raspberry-pi/)
Install ROS Lunar as described here (http://wiki.ros.org/lunar/Installation/Ubuntu)
Compile attached example (udp) and run
  rosrun udp send
  rosrun udp recv

Expected result: recv node prints messages with received numbers.
Actual result: send prints next message
>   Error:   Received an invalid TCPROS header.  Each line must have an equals sign.
>            at line 102 in /tmp/binarydeb/ros-lunar-cpp-common-0.6.9/src/header.cpp

Found in:
Information about system:
  /rosdistro: lunar
  /rosversion: 1.13.6
Ubuntu 16.04(armhf)
ros-lunar-roscpp             1.13.6-0xenial-2018 armhf

The issue was tested using different platforms
ubuntu/armhf/lunar - the issue was found
ubuntu/armhf/jade - the issue was not reproduced
ubuntu/amd64/lunar - the issue was not reproduced
ubuntu/amd64/jade - the issue was not reproduced (upstream packages)

[udp.zip](https://github.com/ros/ros_comm/files/2020600/udp.zip)

**Comment by gavanderhoorn** [comment]
Question on ROS Answers: [UDPROS communication fails at armhf](https://answers.ros.org/question/291076).

Rationale for posting here: if what @AndrijCherepaha describes is correct, this could be a regression.

**Comment by alspitz** [comment]
UDP also does not work for me in Melodic. It works fine on x86_64 but I get

```
Error:   Received an invalid TCPROS header.  Each line must have an equals sign.
         at line 103 in /tmp/binarydeb/ros-melodic-cpp-common-0.6.12/src/header.cpp
```

on an Odroid XU4 (armv7l).

**Comment by alspitz** [comment]
This problem is still here on Melodic.

`Error:   Received an invalid TCPROS header.  Each element must be prepended by a 4-byte length.
         at line 89 in /tmp/binarydeb/ros-melodic-cpp-common-0.6.12/src/header.cpp`

Is anyone working on this or have any tips as to where to start or what may have gone wrong?

This is core ROS functionality completely not working.

**Comment by gavanderhoorn** [comment]
I would assume no one is working on this.

**Comment by kjaget** [comment]
+1 on this - ROS melodic on a Jetson TX2:

rosbag record --udp <topic> 

gives 

Error:   Received an invalid TCPROS header.  Each element must be prepended by a 4-byte length.
         at line 89 in /tmp/binarydeb/ros-melodic-cpp-common-0.6.12/src/header.cpp

**Comment by jveitchmichaelis** [comment]
I'm willing to look at this, as it's completely blocking embedded development for me (on Melodic). The fix to use Kinetic is a workaround, but presumably this will also affect Noetic. This also fails on aarch64, not just armhf.

Running the udp code above I get a different error:

```
[ERROR] [1584487855.546510052]: Header from subscriber did not have the required elements: md5sum, topic, callerid
```

I've cloned the repository and checked out `melodic-devel`. Will try and do some debugging and see what's going on. I'm not familiar with how ROS works at a low level, so I'm assuming that a C++ client will have issues with ros_comm/roscpp and a Python client will call ros_comm/rospy?

If that's the case does the issue occur with a python receiver/subscriber?

**Comment by jveitchmichaelis** [comment]
This does _not_ occur with the publish/subscribe rospy examples. http://wiki.ros.org/rospy/Overview/Publishers%20and%20Subscribers

Some more experiments. I modified @AndrijCherepaha 's example to publish strings. I made four launch files to cover the possible combinations of clients.

### C++ publisher / C++ subscriber:

* `rostopic echo /chatter` prints "hello world" messages
* When launched via rosrun, or with `udp` transport hint: `[ERROR] [1584552392.272356504]: Header from subscriber did not have the required elements: md5sum, topic, callerid`
* When launched with roslaunch, with `unreliable` transport hint: `Error:   Received an invalid TCPROS header.  Each line must have an equals sign.
         at line 103 in /tmp/binarydeb/ros-melodic-cpp-common-0.6.13/src/header.cpp`
* Works if `TransportHints::unreliable()` is removed.

``` Verbose debug log:

[DEBUG] [1584557039.697588286]: /receiver: Publisher update for [/chatter]:  already have these connections:                                                                                                                                                          [15/1866]
Warning: recursive print statement has occurred.  Throwing out recursive print.                                                                                                                                                                                                
process[sender-3]: started with pid [689]                                                                                                                                                                                                                                      
[DEBUG] [1584557040.039609684]: ${node}: remap: __name => sender                                                                                                                                                                                                               
[DEBUG] [1584557040.039786434]: ${node}: remap: __log => /root/.ros/log/6b8562dc-6948-11ea-b0ca-cc4b7326e8f2/sender-3.log                                                                                                                                                      
[DEBUG] [1584557040.056102237]: /sender: Adding tcp socket [7] to pollset                                                                                                                                                                                                      
[DEBUG] [1584557040.056353070]: /sender: UDPROS server listening on port [42167]                                                                                                                                                                                               
[DEBUG] [1584557040.064517096]: /sender: Started node [/sender], pid [689], bound on [herja], xmlrpc port [42874], tcpros port [56321], using [real] time                                                                                                                      
Warning: recursive print statement has occurred.  Throwing out recursive print.                                                                                                                                                                                                
[DEBUG] [1584557040.067191424]: /receiver: Accepted connection on socket [7], new socket [11]                                                                                                                                                                                  
Warning: recursive print statement has occurred.  Throwing out recursive print.                                                                                                                                                                                                
Warning: recursive print statement has occurred.  Throwing out recursive print.                                                                                                                                                                                                
Warning: recursive print statement has occurred.  Throwing out recursive print.                                                                                                                                                                                                
[DEBUG] [1584557040.069585420]: /receiver: Adding tcp socket [11] to pollset
Warning: recursive print statement has occurred.  Throwing out recursive print.
[DEBUG] [1584557040.070146294]: /receiver: TCPROS received a connection from [192.168.86.126:41254]
Warning: recursive print statement has occurred.  Throwing out recursive print.
[DEBUG] [1584557040.071381208]: /receiver: Connection: Creating TransportSubscriberLink for topic [/rosout] connected to [callerid=[/rosout] address=[TCPROS connection on port 50456 to [192.168.86.126:41254 on socket 11]]]
Warning: recursive print statement has occurred.  Throwing out recursive print.
[DEBUG] [1584557040.268894629]: /receiver: Received update for topic [/chatter] (1 publishers)
Warning: recursive print statement has occurred.  Throwing out recursive print.
[DEBUG] [1584557040.269042795]: /receiver: Publisher update for [/chatter]: http://herja:42874/,  already have these connections: 
Warning: recursive print statement has occurred.  Throwing out recursive print.
[DEBUG] [1584557040.270093377]: /receiver: UDPROS server listening on port [48482]
Warning: recursive print statement has occurred.  Throwing out recursive print.
[DEBUG] [1584557040.291556003]: /receiver: Began asynchronous xmlrpc connection to [herja:42874]
Warning: recursive print statement has occurred.  Throwing out recursive print.
Error:   Received an invalid TCPROS header.  Each line must have an equals sign.
         at line 103 in /tmp/binarydeb/ros-melodic-cpp-common-0.6.13/src/header.cpp
[DEBUG] [1584557040.431276116]: /sender: Unable to parse UDPROS connection header: Received an invalid TCPROS header.  Each line must have an equals sign.
```

### Python publisher / Python subscriber

* `rostopic echo /chatter` prints "hello world" messages
*  Messages are received correctly by the subscribing node

### Python publisher / C++ subscriber

* Nothing with `TransportHints().udp` or `TransportHints::unreliable()`
* Works with `TransportHints().reliable` or `TransportHints::tcp()`

### Python subscriber / C++ publisher

* Messages are received correctly by the subscribing node

### Code:
``` python
recv.cc

#include <ros/ros.h>
#include <std_msgs/String.h>

void udpCB (const std_msgs::StringConstPtr& str) {
  ROS_ERROR("Got: [%s]:", str->data.c_str());
}

int main(int argc, char **argv) {
  ros::init(argc, argv, "recv");
  ros::NodeHandle nh;
  ros::Subscriber sub = nh.subscribe(std::string("chatter"), 1, udpCB, ros::TransportHints().unreliable());
  ros::spin();
  return 0;
}

```

``` python
recv_py

#!/usr/bin/python

import rospy
from std_msgs.msg import String

def callback(data):
    rospy.loginfo("Got: %s", data.data)

def listener():
    rospy.init_node('subscriber')
    rospy.Subscriber("chatter", String, callback)
    rospy.spin()

if __name__ == "__main__":
    listener()

```

``` c++
send.cc

#include <ros/ros.h>
#include <std_msgs/String.h>

int main(int argc, char **argv) {
  ros::init(argc, argv, "send");
  ros::NodeHandle nh;
  ros::Publisher pub = nh.advertise<std_msgs::String>(std::string("chatter"), 1, false);
  ros::Rate r(10);

  int i = 0;
  std_msgs::String msg;
  while (ros::ok()) {

    std::stringstream ss;
    ss << "hello world (c++) " << i;
    msg.data = ss.str();

    pub.publish(msg);
    ++i;
    r.sleep();
  }

  return 0;
}

```

``` python
send_py

import rospy
from std_msgs.msg import String

pub = rospy.Publisher('chatter', String, queue_size=10)
rospy.init_node('publisher')
r = rospy.Rate(10) # 10hz

while not rospy.is_shutdown():
   pub.publish("hello world")
   r.sleep()


```

**Comment by jveitchmichaelis** [comment]
* Overlaying `ros_comm:kinetic-devel` inside a ROS `melodic` install works as expected, no errors with unreliable transport.

* Using `melodic-devel` and checking out `kinetic-devel` specifically for `utilities/xmlrpcpp` works.

* The libxmlrpc tests in `melodic-devel` fail on ARM with `-WNarrowing`. And this _is_ something that changed between kinetic and lunar after this commit: https://github.com/ros/ros_comm/commit/7019fb1604363c5ca1849a88930202c7ce28aed6#diff-abead072a4353c6c959717c854289c83

Output from `catkin build test_rospy --no-deps --catkin-make-args run_tests`:

```
Errors     << xmlrpcpp:make /ws/logs/xmlrpcpp/build.make.001.log
In file included from /usr/src/googletest/googletest/include/gtest/gtest.h:62:0,
                 from /ws/src/ros_comm/utilities/xmlrpcpp/test/test_base64.cpp:27:
/ws/src/ros_comm/utilities/xmlrpcpp/test/test_base64.cpp: In function ‘testing::internal::ParamGenerator<Base64TestData> gtest_MultilineBase64Test_EvalGenerator_()’:
/ws/src/ros_comm/utilities/xmlrpcpp/test/test_base64.cpp:107:39: error: narrowing conversion of ‘-128’ from ‘int’ to ‘char’ inside { } [-Wnarrowing]
             "Hy8/T19vf4+fr7/P3+/w==\n")));
                                       ^
/usr/src/googletest/googletest/include/gtest/gtest-param-test.h:1423:66: note: in definition of macro ‘INSTANTIATE_TEST_CASE_P’
       gtest_##prefix##test_case_name##_EvalGenerator_() { return generator; } \
                                                                  ^~~~~~~~~
make[3]: *** [test/CMakeFiles/test_base64.dir/test_base64.cpp.o] Error 1
make[2]: *** [test/CMakeFiles/test_base64.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/run_tests.dir/rule] Error 2
make: *** [run_tests] Error 2
cd /ws/build/xmlrpcpp; catkin build --get-env xmlrpcpp | catkin env -si  /usr/bin/make run_tests --jobserver-fds=6,7 -j; cd -
```

Edit: After all that...

Aside from these tests failling, this should have been fixed in https://github.com/ros/ros_comm/commit/61f0b767ade6b567b1bc8870ef4ecaca355830c7. 

This seems to have been fixed with the latest version from `apt` (`1.14.4`). The Docker images may still be out of date at time of writing.

@dirk-thomas This issue can be closed I think. I've raised a separate issue for the tests: https://github.com/ros/ros_comm/issues/1912

**Comment by jveitchmichaelis** [comment]
Ah well after all that I believe this has been fixed recently? 

https://github.com/ros/ros_comm/issues/1768
